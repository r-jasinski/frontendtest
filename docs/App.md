# App.vue

This is the main Vue component of the application. It is responsible for rendering the game board, players, and sidebar.

## Template Structure

The template is divided into two main sections:

1. The first section contains the game board and player details. It uses the `Player` and `Board` components. Each `Player` component receives props for flag, image, name, and time. The `Board` component receives a prop for handling square selection. There's also a settings icon rendered using `GearsSVG`.

2. The second section contains the `Sidebar` component which receives notations as a prop.

## Script

The script imports the necessary components (`Board`, `Player`, `GearsSVG`, `Sidebar`) and create a type `Notation`.

## Props

- `Player` component:

  - `flag`: A string representing the path to the flag image.
  - `image`: A string representing the path to the player's image.
  - `name`: A string representing the player's name.
  - `time`: A string representing the player's time.
  - `is-player-turn`: A boolean indicating if it's this player's turn.

- `Board` component:

  - `on-square-select`: A function for handling the event when a square is selected.

- `Sidebar` component:
  - `notations`: An array of `Notation` objects.
