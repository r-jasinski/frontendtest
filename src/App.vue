<template>
  <main>
    <section>
      <div class="game">
        <PlayerTag
          flag="images/flag.png "
          image="images/black_400.png"
          name="Guest2939666985"
          time="5:00"
        />
        <ChessBoard @squareSelect="handleSelect" />
        <PlayerTag
          flag="images/flag.png "
          image="images/white_400.png"
          is-player-turn
          name="Guest9241962816"
          time="4:58"
        />
        <span class="settings">
          <GearsSVG />
        </span>
      </div>
    </section>
    <section>
      <SideBar :notations="notations" />
    </section>
  </main>
</template>

<script lang="ts">
import ChessBoard from './components/ChessBoard.vue'
import GearsSVG from './assets/icons/gears.svg?component'
import PlayerTag from './components/PlayerTag.vue'
import SideBar from './components/SideBar.vue'

export type Notation = {
  square: number
  notation: string
}

export default {
  name: 'App',
  components: { ChessBoard, PlayerTag, GearsSVG, SideBar },
  data() {
    return {
      notations: [] as Notation[]
    }
  },
  methods: {
    /**
     * Handles the selection of a square on the game board.
     *
     * @param {number} square - The number of the square that was selected.
     * The squares are numbered from 1 to 64, starting from the top left of the board and going right and down.
     *
     * This method calculates the file (column) and rank (row) of the selected square in standard chess notation.
     * The file is a letter from 'a' to 'h' and the rank is a number from 1 to 8.
     * It then creates a Notation object with the square number and its notation, and adds it to the notations array.
     */
    handleSelect(square: number) {
      const file = String.fromCharCode(104 - ((square - 1) % 8))
      const rank = Math.ceil(square / 8)
      const data: Notation = { square, notation: `${file}${rank}` }
      this.notations.push(data)
    }
  }
}
</script>

<style scoped>
main {
  height: 100vh;
  width: 100vw;
  min-width: 340px;
  display: flex;
  place-items: center;
  padding: 2rem;
  gap: 34px;
}
section {
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  justify-content: flex-start;
}
.game {
  height: 80%;
  aspect-ratio: 1;
  display: flex;
  flex-direction: column;
  gap: 10px;
  position: relative;
}
.settings {
  position: absolute;
  right: 0;
  translate: 24px;
  opacity: 0.5;
}
@media screen and (max-width: 768px) {
  main {
    padding: 0;
    flex-direction: column;
  }
  .game {
    width: 100%;
    margin-bottom: 16px;
  }
  .settings {
    right: 8px;
    bottom: 0;
    translate: 0 128px;
    z-index: 999;
  }
}
</style>
