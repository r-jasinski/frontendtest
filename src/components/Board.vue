<template>
  <div class="board" @keydown.prevent tabindex="0">
    <span
      v-for="i in Array.from({ length: 64 }, (_, i) => 64 - i)"
      :data-file="getDataFile(i)"
      :data-rank="getDataRank(i)"
      :key="i"
      @click="onSquareSelect(i)"
      tabindex="1"
    >
    </span>
    <GearsSVG />
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'Board',
  props: { onSquareSelect: { type: Function, required: true } },
  methods: {
    getDataRank(square: number) {
      const rank = Math.ceil(square / 8)
      return square % 8 === 0 ? rank : null
    },
    getDataFile(square: number) {
      return square <= 8 ? String.fromCharCode(104 - ((square - 1) % 8)) : null
    }
  }
})
</script>

<style scoped>
.board {
  height: 100%;
  aspect-ratio: 1;
  position: relative;
  display: grid;
  grid-template-columns: repeat(8, 1fr);
  clip-path: inset(0 0 0 0 round 4px);
  background-color: var(--color-gray-300);
  color: var(--color-green-300);
}
span:nth-child(1) ~ span:nth-child(-2n + 8),
span:nth-child(8) ~ span:nth-child(-2n + 15),
span:nth-child(16) ~ span:nth-child(-2n + 24),
span:nth-child(24) ~ span:nth-child(-2n + 31),
span:nth-child(32) ~ span:nth-child(-2n + 40),
span:nth-child(40) ~ span:nth-child(-2n + 47),
span:nth-child(48) ~ span:nth-child(-2n + 56),
span:nth-child(56) ~ span:nth-child(-2n + 65) {
  background-color: var(--color-green-300);
  color: var(--color-gray-300);
}
span {
  width: 100%;
  height: 100%;
  aspect-ratio: 1;
  position: relative;
  outline: none;
  border: none;
  font-size: 12px;
  font-weight: bold;
  cursor: pointer;
}
span:focus:after {
  content: '';
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: rgb(255, 255, 51);
  opacity: 0.5;
}
span[data-rank]:focus:after {
  width: calc(100% - 4px);
  height: calc(100% - 4px);
}
span[data-rank]:after {
  content: attr(data-rank);
  position: absolute;
  padding-left: 4px;
  padding-top: 4px;
}
span[data-file]:before {
  content: attr(data-file);
  position: absolute;
  bottom: 0;
  right: 0;
  margin: 4px;
}
</style>
