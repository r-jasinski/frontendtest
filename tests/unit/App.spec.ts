import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import App from '../../src/App.vue'
import Board from '../../src/components/ChessBoard.vue'

describe('App.vue', () => {
  const wrapper = mount(App)
  it('should render children components', () => {
    const board = wrapper.findComponent(Board)
    expect(board.exists()).toBe(true)
  })
  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
