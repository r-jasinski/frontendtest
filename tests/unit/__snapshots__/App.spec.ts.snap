// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App.vue > should match snapshot 1`] = `
<main
  data-v-7a7a37b1=""
>
  <section
    data-v-7a7a37b1=""
  >
    <div
      class="game"
      data-v-7a7a37b1=""
    >
      <div
        class="player"
        data-v-7a7a37b1=""
        data-v-a6c9493b=""
      >
        <div
          class="tagline"
          data-v-a6c9493b=""
        >
          <img
            alt="Player image"
            data-v-a6c9493b=""
            src="images/black_400.png"
          />
          <div
            class="user-tagline"
            data-v-a6c9493b=""
          >
            <span
              data-v-a6c9493b=""
            >
              Guest2939666985
            </span>
            <img
              alt="Player flag"
              data-v-a6c9493b=""
              src="images/flag.png "
            />
            <ul
              data-v-a6c9493b=""
            >
              <li
                data-v-a6c9493b=""
              />
              <li
                data-v-a6c9493b=""
              />
              <li
                data-v-a6c9493b=""
              />
              <li
                data-v-a6c9493b=""
              />
            </ul>
          </div>
        </div>
        <div
          class="clock inactive"
          data-v-a6c9493b=""
        >
          </src/assets/icons/clock.svg?component
            data-v-a6c9493b=""
          />
          <span
            data-v-a6c9493b=""
          >
            5:00
          </span>
        </div>
      </div>
      <div
        class="board"
        data-v-41886935=""
        data-v-7a7a37b1=""
        tabindex="0"
      >
        
        <span
          data-rank="8"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-rank="7"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-rank="6"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-rank="5"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-rank="4"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-rank="3"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-rank="2"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-file="a"
          data-rank="1"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-file="b"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-file="c"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-file="d"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-file="e"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-file="f"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-file="g"
          data-v-41886935=""
          tabindex="1"
        />
        <span
          data-file="h"
          data-v-41886935=""
          tabindex="1"
        />
        
      </div>
      <div
        class="player"
        data-v-7a7a37b1=""
        data-v-a6c9493b=""
      >
        <div
          class="tagline"
          data-v-a6c9493b=""
        >
          <img
            alt="Player image"
            data-v-a6c9493b=""
            src="images/white_400.png"
          />
          <div
            class="user-tagline"
            data-v-a6c9493b=""
          >
            <span
              data-v-a6c9493b=""
            >
              Guest9241962816
            </span>
            <img
              alt="Player flag"
              data-v-a6c9493b=""
              src="images/flag.png "
            />
            <ul
              data-v-a6c9493b=""
            >
              <li
                data-v-a6c9493b=""
              />
              <li
                data-v-a6c9493b=""
              />
              <li
                data-v-a6c9493b=""
              />
              <li
                data-v-a6c9493b=""
              />
            </ul>
          </div>
        </div>
        <div
          class="clock"
          data-v-a6c9493b=""
        >
          </src/assets/icons/clock.svg?component
            data-v-a6c9493b=""
          />
          <span
            data-v-a6c9493b=""
          >
            4:58
          </span>
        </div>
      </div>
      <span
        class="settings"
        data-v-7a7a37b1=""
      >
        </src/assets/icons/gears.svg?component
          data-v-7a7a37b1=""
        />
      </span>
    </div>
  </section>
  <section
    data-v-7a7a37b1=""
  >
    <div
      class="sidebar"
      data-v-7a7a37b1=""
      data-v-ce868814=""
    >
      <div
        class="tabs"
        data-v-ce868814=""
      >
        <span
          data-v-ce868814=""
        >
          </src/assets/icons/timer.svg?component
            data-v-ce868814=""
          />
           Play 
        </span>
        <span
          data-v-ce868814=""
        >
          </src/assets/icons/more.svg?component
            data-v-ce868814=""
          />
           New Game 
        </span>
        <span
          data-v-ce868814=""
        >
          </src/assets/icons/board.svg?component
            data-v-ce868814=""
          />
           Games 
        </span>
        <span
          data-v-ce868814=""
        >
          </src/assets/icons/players.svg?component
            data-v-ce868814=""
          />
           Players 
        </span>
      </div>
      <div
        class="opening"
        data-v-ce868814=""
      >
         Center Game: Beyer Gambit 
        <span
          data-v-ce868814=""
        >
          </src/assets/icons/book.svg?component
            data-v-ce868814=""
          />
          </src/assets/icons/compass.svg?component
            data-v-ce868814=""
          />
        </span>
      </div>
      <div
        class="move-list"
        data-v-ce868814=""
      >
        <ol
          data-v-ce868814=""
        >
          
          
          <li
            data-v-ce868814=""
          />
        </ol>
      </div>
      <div
        class="game-buttons"
        data-v-ce868814=""
      >
        <span
          data-v-ce868814=""
        >
          </src/assets/icons/half.svg?component
            data-v-ce868814=""
          />
          Draw
          </src/assets/icons/flag.svg?component
            data-v-ce868814=""
          />
          Resign
        </span>
        <span
          data-v-ce868814=""
        >
          </src/assets/icons/first.svg?component
            data-v-ce868814=""
          />
          </src/assets/icons/previous.svg?component
            data-v-ce868814=""
          />
          </src/assets/icons/next.svg?component
            data-v-ce868814=""
          />
          </src/assets/icons/last.svg?component
            data-v-ce868814=""
          />
          </src/assets/icons/flip.svg?component
            data-v-ce868814=""
          />
        </span>
      </div>
      <div
        class="chat"
        data-v-ce868814=""
      >
        <span
          data-v-ce868814=""
        >
          NEW GAME
        </span>
        <span
          data-v-ce868814=""
        >
          Guest2939666985 vs Guest9241962816 (5 min Unrated)
        </span>
      </div>
      <div
        class="chat-login"
        data-v-ce868814=""
      >
         Please 
        <a
          data-v-ce868814=""
          href="#"
        >
          Sign Up
        </a>
         to use the chat. 
      </div>
    </div>
  </section>
</main>
`;
